### PR TITLE
Feature/2826 smartystreets verified

### LINKED ISSUE #2826: SmartyStreets not properly setting Verified checkbox when verification is attempted but fails.

### PR DESCRIPTION
**Pull request by njjc**
# Critical Changes

# Changes

# Issues Closed
- Fixes #2826

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: src/classes/ADDR_SmartyStreets_Validator.cls [hunk 1]
```diff
@@ -143,69 +143,70 @@ public with sharing class ADDR_SmartyStreets_Validator implements ADDR_IValidato
 
         //Don't even make the address verification request to SmartyStreets if no US address is included, to save in API calls
         if(includesUSAddress) {
```
Comment (randi274) [inline src/classes/ADDR_SmartyStreets_Validator.cls]: If a US Address isn't included, it doesn't set verified on the address. It could be verified and set as a foreign address.

File: src/classes/ADDR_SmartyStreets_Validator.cls [hunk 2]
```diff
@@ -222,66 +223,68 @@ public with sharing class ADDR_SmartyStreets_Validator implements ADDR_IValidato
         
         //Don't even make the address verification request to SmartyStreets if no US address is included, to save in API calls
         if(includesUSAddress) {
```
Comment (randi274) [inline src/classes/ADDR_SmartyStreets_Validator.cls]: Same comment re: US Address.

File: src/classes/ADDR_SmartyStreets_Validator.cls [hunk 3]
```diff
@@ -222,66 +223,68 @@ public with sharing class ADDR_SmartyStreets_Validator implements ADDR_IValidato
...
-	                av.Verification_Status__c = Label.Zip_Not_Found;
-	                ERR_Handler.processError(ex, ERR_Handler_API.Context.ADDR);
-	            }
-	            ERR_Handler.processError(ex, ERR_Handler_API.Context.ADDR);
-	        }
+            try {
+                ADDR_SmartyStreets_Gateway la = new ADDR_SmartyStreets_Gateway(settings);
+                List<ADDR_SmartyStreets_Gateway.ZipResponse> responses = la.verifyZips(zrList);
+                UTIL_Debug.debug('****Gateway responses: ' + JSON.serializePretty(responses));
+                if (responses != null && !responses.isEmpty()) {
```
Comment (randi274) [inline src/classes/ADDR_SmartyStreets_Validator.cls]: Why are we checking for !responses.isEmpty() here but not above? It looks like we're doing the empty check already in verifyAddresses and verifyZips.

File: src/classes/ADDR_SmartyStreets_Validator.cls [hunk 4]
```diff
@@ -143,69 +143,70 @@ public with sharing class ADDR_SmartyStreets_Validator implements ADDR_IValidato
...
+                            av.Ambiguous__c = true;
+                            responseGroup.add(sr);
+                        }
+                        String response = JSON.serializePretty(responseGroup);
+                        av.API_Response__c = trimResponse(response);
+                        av.Verification_Status__c = Label.Addr_Verified;
+                    }
+                }
+                for (Address__c av : avList) {             
+                    if (av.API_Response__c == null && !String.isBlank(settings.Auth_ID__c) &&
```
Comment (randi274) [inline src/classes/ADDR_SmartyStreets_Validator.cls]: Is this check on the Auth ID, Auth Token and Class a little late? Should we be doing this check at an earlier point in the code? If not, why is this required here, but not in verifyZipRecords?

File: src/classes/ADDR_SmartyStreets_Validator.cls (hunk 1 above)
Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Good question. This is feels like a spec issue that we can discuss with @mpusto tomorrow.

File: src/classes/ADDR_SmartyStreets_Validator.cls (hunk 4 above)
Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Agreed that this is a really weird place for this check. I'll have to dig in a bit to figure out if we should move it up or remove it altogether.

File: src/classes/ADDR_SmartyStreets_Validator.cls (hunk 3 above)
Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Right, I think we can remove the check since VerifyZips does it already and returns null.

File: src/classes/ADDR_SmartyStreets_Validator.cls [hunk 5]
```diff
@@ -222,66 +223,68 @@ public with sharing class ADDR_SmartyStreets_Validator implements ADDR_IValidato
...
+                List<ADDR_SmartyStreets_Gateway.ZipResponse> responses = la.verifyZips(zrList);
+                UTIL_Debug.debug('****Gateway responses: ' + JSON.serializePretty(responses));
+                if (responses != null && !responses.isEmpty()) {
+                    if(responses.size() == avList.size()) {
+                        for (integer i = 0; i < responses.size(); i++) {
+                            ADDR_SmartyStreets_Gateway.ZipResponse sr = responses[i];
+                            Address__c av = avList[i];             
+        
+                            // look for an error
+                            if (sr.status != null && sr.reason != null) {
```
[the remark this answers is not in the document]
Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Good catch, I'll set it.

File: src/classes/ADDR_SmartyStreets_Validator.cls (hunk 4 above)
Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Yep, we can remove these checks since we'll only get to this code path if the class setting is set to this class, and we're already checking if the auth vars exist on the ADDR_SmartyStreets_Gateway lines [54](https://github.com/SalesforceFoundation/Cumulus/blob/master/src/classes/ADDR_SmartyStreets_Gateway.cls#L54) and [185](https://github.com/SalesforceFoundation/Cumulus/blob/master/src/classes/ADDR_SmartyStreets_Gateway.cls#L185). I'll remove it.

File: src/classes/ADDR_SmartyStreets_Validator.cls (hunk 1 above)
Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: @mpusto Randi noticed that our SmartyStreets address verification code filters out non-US addresses. It seems that when we implemented the feature years ago, they only supported US address verification, but in visiting smartystreets.com it seems like they now support international addresses. I think we could use a work item to investigate updating our code to stop filtering out international addresses with SmartyStreets, what do you think?

Comment (mpusto) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Yes, I agree! It's come up in the HUB before and I answered with "well they didn't support it when we wrote the code". :)

Comment (njjc) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: Perfect, I'll let you add it to the backlog! Thanks.

Comment (mpusto) [reply src/classes/ADDR_SmartyStreets_Validator.cls]: https://github.com/SalesforceFoundation/Cumulus/issues/2853
